### ISSUE TITLE
Enhancements around proxy and hostname configuration

### ISSUE THREAD

**Issue by stianst** [keycloak/keycloak#15018]
### Description

Based on all the reported issues around configuring hostname/proxy we do still have some usability or documentation improvements that can be made.

It's worth going through the issues linked in https://github.com/keycloak/keycloak/issues/14666 to get a better understanding with what folks are struggling with.

**Comment by stianst** [comment]
Many people are facing these issues when loading the admin console. Could we somehow improve the error messages or output the user gets in these situations?

**Comment by stianst** [comment]
The documentation doesn't really talk all that much about how to check the configuration around proxy/hostname. Maybe we could even have a "test" endpoint of sorts that make it easy to discover what the URLs results in for a given configuration?

**Comment by stianst** [comment]
Documentation doesn't mention clearly that Keycloak can only be exposed through a single frontend URL, and a single admin URL. This should be made clear, and we should explain what folks can do if they really need to use multiple URLs (separate instances, enforce host headers etc on proxy, custom hostname provider).

**Comment by stianst** [comment]
Maybe we can make the proxy/hostname configuration even more use-case driven?

**Comment by stianst** [comment]
Proxy documentation mentions X-Forwarded-* headers, but not when they are used/not-used. X-Forwarded-For should always be needed as long as there's a proxy, but X-Forwaded-Scheme, X-Forwarded-Hostname, and X-Forwaded-Port should maybe just be excluded completely when there is hostname provider configured correctly?

**Comment by stianst** [comment]
@pedroigor FIY

**Comment by stianst** [comment]
Here's a case where someone is trying to access the admin URL over HTTP, while having HTTPS enabled:

https://github.com/keycloak/keycloak/issues/14812

That'll boil down to redirect-uris/web-origin for the admin console being configured for HTTPS, which is why there's a 403 when loading the session iframe.

I don't think this one is something we should support. Why go to the trouble of configuring HTTPS, but then not use it?

We should document this though perhaps?

**Comment by pedroigor** [comment]
We used to have the steps to enable HTTP documented but we decided to remove the text from the guide :) See https://github.com/keycloak/keycloak/pull/13681#discussion_r953561225.

**Comment by pedroigor** [comment]
I think we can consider:

* Keep only `hostname`, `hostname-admin`, `hostname-url` and `hostname-admin-url`
* Remove all proxy modes and use `proxy-enabled` to enable forward headers

By doing this I would expect confusion around hostname and proxy by reducing the number of options they can set.

**Comment by stianst** [comment]
I've written a util to help debug hostname issues. 

Download https://github.com/stianst/playground/releases/download/1.0.0/hostname-debug.jar and add to the providers directory.

You can then open the tool with `KEYCLOAK_BASE_URL/realms/REALM/hostname-debug` for example `http://localhost:8080/realms/master/hostname-debug`. It should be opened from a browser, and not curl for example.

The output can also be copy/pasted to a GitHub issue to help figure out what someone is struggling with.

**Comment by stianst** [comment]
Discovered an issue with `hostname-strict=false`, where even with that set it will still force `https` in production:
https://github.com/keycloak/keycloak/issues/15287

**Comment by grossws** [comment]
I had to tune `Content-Security-Policy` header to allow admin.v2 with different hostnames.

Configuration:
* `KC_HOSTNAME=https://sso.example.org`
* `KC_HOSTNAME_ADMIN=https://keycloak-admin.internal.example.org`
* `KC_PROXY=reencrypt`
* nginx proxy with `keycloak-admin` reachable only through VPN on private ip, `/admin/` is only exposed for that internal address

`CSP` header: `frame-src 'self' https://sso.example.org; frame-ancestors 'self' https://keycloak-admin.internal.example.org; object-src 'none';`

Otherwise both Chrome and Firefox block 3rdparty cookie check iframe and prevent loading console with following error
```
Uncaught (in promise) {error: 'Timeout when waiting for 3rd party check iframe message.'}
```

Without external URL in the `frame-src`: 
```
Refused to frame 'https://sso.example.org/' because it violates the following Content Security Policy directive: "frame-src 'self'".
```

Without internal URL in the `frame-ancestors`: 
```
Refused to frame 'https://sso.example.org/' because an ancestor violates the following Content Security Policy directive: "frame-ancestors 'self'".
```

**Comment by pedroigor** [comment]
@grossws We should probably do that by default when configuring the CSP header. In this case, we would also take into account the `hostname-admin-url`.

**Comment by vmuzikar** [comment]
We're investigating a possible [FAQ section in the docs](https://github.com/keycloak/keycloak/issues/16018). It might help resolving this issue.

**Comment by ruckc** [comment]
The best solution to solving this is simplifying the implementation.  Ideally, just a single variable that makes keycloak leverage the Host/X-Forwarded headers provided by a reverse proxy across the board.  A second variable to control if admin should be exposed.  

Keycloak acts unintuitively when needing to configure URLs.  Most modern applications honor Host headers or X-Forwarded headers out of the box.  Adding documentation around this doesn't help the inane implementation.  Quarkus natively can handle these headers out of the box, but Keycloak's current implementation chooses not to leverage this capability and implemented its own confusing, unique, and cumbersome implementation.

**Comment by pedroigor** [comment]
@ruckc Thanks for your feedback. 

One of the main reasons why we have the hostname settings is that the server must be capable of building responses and content (e.g.: emails) using a fixed URL. We also allow setting this URL on a per-realm basis. It is not only about honoring the headers set by the proxy ...

**Comment by stianst** [comment]
> 

In addition there are host header injection (Host or X-Forwarded-Host) attacks that opens up vulnerabilities if Keycloak blindly trusts the hostnamed from headers.

**Comment by ruckc** [comment]
I understand the host injection... But only idiots run application servers
without a proper reverse proxy, which is where the current status quo runs
into issues

**Comment by stianst** [comment]
> 

That's a completely incorrect statement. Several smaller deployments do not have a reverse proxy in front of it, many people don't have control over the reverse proxy, and it is also very common that the Host or X-Forwarded-Host headers are blindly set by the reverse proxy without any validation.

As the docs state if you are using a reverse proxy that can guarantee a valid X-Forwarded-Host header you can disable the strict hostname checking, and it will infer it from the headers instead.

**Comment by slominskir** [comment]
One usability issue is that the default configuration restricts the admin console to the same origin as the front hostname.  So if you do everything that is recommended to setup Keycloak the admin console doesn't work immediately due to CORS / web origins.  For example, the recommended approach is to:
1. Use two or more instances of Keycloak behind a proxy/load balancer
2. Don't expose the admin console through the proxy

So the easiest way is on each Keycloak node set hostname-admin to the hostname of first the keycloak node.  However, you also must update the webOrigins of the security-admin-console client of the master realm.  And if you've now just broke the web admin console you can't use it anymore without undoing some configuration and exposing the web console back through the proxy so you may end up messing with CORS via CLI:

```
${KEYCLOAK_HOME}/bin/kcadm.sh config credentials --server "${KEYCLOAK_SERVER_URL}" --realm master --user "${KEYCLOAK_ADMIN}" --password "${KEYCLOAK_ADMIN_PASSWORD}"

CLIENT_ID=`${KEYCLOAK_HOME}/bin/kcadm.sh get clients -r master --fields clientId,id | jq -r '.[] | select(.clientId=="security-admin-console") | .id'`

${KEYCLOAK_HOME}/bin/kcadm.sh update clients/${CLIENT_ID} -r master -s webOrigins=["+"]
```

Maybe the default should be "+"?

Alternatively, this could be handled as a special case - even though the admin console is accessed directly on a backend node as recommended, the authorization still happens on the frontend and this is where you see issues with the login-status-iframe and CORS (so this may actually be a bug?).

Also consider adding official Docker Compose projects to accompany the existing guides.  Probably want to have different compose for a few different load balancers (nginx, haproxy).  I've got an example [here](https://github.com/slominskir/keycloak_cluster/blob/main/haproxy/haproxy.cfg), though it's currently doesn't match what I have in production because it doesn't hide the admin console on a backend node or use sticky sessions.

**Comment by slominskir** [comment]
The example default/recommended haproxy config should probably be more like:
```
frontend front
...
  use_backend auth if { path_beg /auth/realms/ /auth/resources/ }

backend auth
  balance roundrobin
  cookie KC_ROUTE insert indirect nocache
  server keycloak1 keycloak1.example.com:443 ssl ca-file /etc/pki/ca-bundle.crt check cookie keycloak1
  server keycloak2 keycloak2.example.com:443 ssl ca-file /etc/pki/ca-bundle.crt check cookie keycloak2
```

**Comment by pedroigor** [comment]
@slominskir The problem with adding proxy configuration in our docs is that we need to maintain it as they change. I usually use HAProxy to test HA deployments and, although my configuration is the same for quite some time, it happens to change.

I think there is room for improvement when exposing the admin console using a separate URL. However, IIRC, we force CSP policies when responding with the admin console GET request. But we do not set CORS.

**Comment by stianst** [comment]
> One usability issue is that the default configuration restricts the admin console to the same origin as the front hostname.  So if you do everything that is recommended to setup Keycloak the admin console doesn't work immediately due to CORS / web origins.  For example, the recommended approach is to:
> 1. Use two or more instances of Keycloak behind a proxy/load balancer
> 2. Don't expose the admin console through the proxy
> 
> So the easiest way is on each Keycloak node set hostname-admin to the hostname of first the keycloak node.  However, you also must update the webOrigins of the security-admin-console client of the master realm.  And if you've now just broke the web admin console you can't use it anymore without undoing some configuration and exposing the web console back through the proxy so you may end up messing with CORS via CLI:
> 
> ```
> ${KEYCLOAK_HOME}/bin/kcadm.sh config credentials --server "${KEYCLOAK_SERVER_URL}" --realm master --user "${KEYCLOAK_ADMIN}" --password "${KEYCLOAK_ADMIN_PASSWORD}"
> 
> CLIENT_ID=`${KEYCLOAK_HOME}/bin/kcadm.sh get clients -r master --fields clientId,id | jq -r '.[] | select(.clientId=="security-admin-console") | .id'`
> 
> ${KEYCLOAK_HOME}/bin/kcadm.sh update clients/${CLIENT_ID} -r master -s webOrigins=["+"]
> ```
> 
> Maybe the default should be "+"?
> 
> Alternatively, this could be handled as a special case - even though the admin console is accessed directly on a backend node as recommended, the authorization still happens on the frontend and this is where you see issues with the login-status-iframe and CORS (so this may actually be a bug?).
> 
> Also consider adding official Docker Compose projects to accompany the existing guides.  Probably want to have different compose for a few different load balancers (nginx, haproxy).  I've got an example [here](https://github.com/slominskir/keycloak_cluster/blob/main/haproxy/haproxy.cfg), though it's currently doesn't match what I have in production because it doesn't hide the admin console on a backend node or use sticky sessions.

The cors origin uses the admin URL already (or at least is supposed to). If you're getting cors issues it's most likely due to the URLs being wrong.

**Comment by slominskir** [comment]
> The cors origin uses the admin URL already (or at least is supposed to). If you're getting cors issues it's most likely due to the URLs being wrong.

It's quite possible I did something wrong, but I think all I did was update `keycloak.conf` with:
```
hostname=haproxy.example.com
hostname-admin=keycloak1.example.com
```

This was done on both keycloak1 and keycloak2 nodes (and then I ran `kc.sh build` before starting).  Then I was surprised to discover navigating to the admin console at `https://keycloak1.example.com/auth/admin/master/console` results in an XHR request that fails with 403 and the admin screen hangs with "Loading the admin console" and spin graphic.  The XHR request that fails is: 
```
https://haproxy.example.com/auth/realms/master/protocol/openid-connect/login-status-iframe.html/init?client_id=security-admin-console&origin=https://keycloak1.example.com
```
I expected this would just work given I explicitly set the `hostname` and `hostname-admin`, but to get this to work I had to add `+` as the webOrigins of the `security-admin-console` client of the `master` realm.

I also noticed that in the admin console you can't delete the webOrigins "+" as it fails with "client already exists" on save.   That's probably a separate UI bug.  You can replace it with a space though and save that.  If you unset webOrigins and then try another operation that fetches a token you see a more pointed error in the browser developer console:

```
Access to XMLHttpRequest at 'https://haproxy.example.com/auth/realms/master/protocol/openid-connect/token' from origin 'https://keycloak1.example.com' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource.
```

**Comment by pedroigor** [comment]
@slominskir Can you link the full HAProxy configuration you are using? I would like to try it out and see what can be improved.

**Comment by slominskir** [comment]
Using RHEL8 yum installed haproxy version 1.8.27:

```
global
  log 127.0.0.1 local2
  chroot  /var/lib/haproxy
  pidfile /run/haproxy.pid
  maxconn 4000
  user haproxy
  group haproxy
  daemon
  stats socket /var/lib/haproxy/stats
  ssl-default-bind-ciphers PROFILE=SYSTEM
  ssl-default-server-ciphers PROFILE=SYSTEM

defaults
  mode http
  log global
  option httplog
  option dontlognull
  timeout client 10s
  timeout connect 5s
  timeout server 10s
  timeout http-request 10s
  timeout queue 1m
  timeout check 10s
  timeout http-keep-alive 10s

frontend front
  bind x.x.x.x:443 ssl crt /etc/pki/tls/private/haproxy.pem alpn h2,http/1.1
  option forwardfor
  http-request set-header X-Forwarded-Proto https
  http-request set-header X-Forwarded-Host %[req.hdr(Host)]
  use_backend auth if { path_beg /auth/realms/ /auth/resources/ }

backend auth
  balance roundrobin
  cookie KC_ROUTE insert indirect nocache
  server keycloak1 keycloak1.example.com:443 ssl ca-file /etc/pki/tls/certs/ca-bundle.crt check cookie keycloak1
  server keycloak2 keycloak2.example.com:443 ssl ca-file /etc/pki/tls/certs/ca-bundle.crt check cookie keycloak2
```

**Comment by pedroigor** [comment]
Here is what I tried:

```
#---------------------------------------------------------------------
# Global settings
#---------------------------------------------------------------------
global
    # to have these messages end up in /var/log/haproxy.log you will
    # need to:
    #
    # 1) configure syslog to accept network log events.  This is done
    #    by adding the '-r' option to the SYSLOGD_OPTIONS in
    #    /etc/sysconfig/syslog
    #
    # 2) configure local2 events to go to the /var/log/haproxy.log
    #   file. A line like the following can be added to
    #   /etc/sysconfig/syslog
    #
    #    local2.*                       /var/log/haproxy.log
    #
    log         127.0.0.1 local2

    chroot      /var/lib/haproxy
    pidfile     /var/run/haproxy.pid
    maxconn     4000
    user        haproxy
    group       haproxy
    daemon

    # turn on stats unix socket
    stats socket /var/lib/haproxy/stats

    # utilize system-wide crypto-policies
    ssl-default-bind-ciphers PROFILE=SYSTEM
    ssl-default-server-ciphers PROFILE=SYSTEM

#---------------------------------------------------------------------
# common defaults that all the 'listen' and 'backend' sections will
# use if not designated in their block
#---------------------------------------------------------------------
defaults
    mode                    http
    log                     global
    option                  httplog
    option                  dontlognull
    option http-server-close
    option forwardfor       except 127.0.0.0/8
    option                  redispatch
    retries                 3
    timeout http-request    10s
    timeout queue           1m
    timeout connect         10s
    timeout client          1m
    timeout server          1m
    timeout http-keep-alive 10s
    timeout check           10s
    maxconn                 3000

frontend mykeycloaktls
    # Copy the haproxy.crt.pem file to /etc/haproxy
    bind *:443 ssl crt /etc/haproxy/haproxy.crt.pem
    use_backend mykeycloaktls

backend mykeycloaktls
    stats enable
    stats uri /haproxy?status
    option httpchk
    http-check send meth GET uri /health
    http-check expect status 200
    http-request redirect scheme https unless { ssl_fc }
    http-request set-header X-Forwarded-Proto https
    http-request set-header X-Forwarded-Port 443
    http-request set-header X-Forwarded-For %[src]
    cookie KC_ROUTE insert indirect nocache
    balance roundrobin
    server kc1 127.0.0.1:8443 ssl verify none
    server kc2 127.0.0.1:8543 ssl verify none
```

Then start both instances using:

```
kc.sh start --https-key-store-file=<my_keystore> --https-key-store-password=<my_keystore_secret> --http-port=<port> --https-port=<port> --health-enabled true --hostname mykeycloak --hostname-admin admin.mykeycloak --proxy reencrypt
```

Admin console is loading fine.

**Comment by slominskir** [comment]
Ideally we'd have a working example project on GitHub (preferably using Docker compose) such that we can run it and tweak it to see what differences may be causing the behavior I'm seeing.  Docker helps minimize the "works for me problem".  The config you posted is decidedly different than what I posted and also many of the more subtle things could cause the undesired behavior I'm seeing.   For example, I use the keycloak config file, not the command line to config Keycloak.   Also, I never set the http-port config (only https).  Exact versions of Keycloak and HAProxy matter.   Lots of little things like that and one of them probably tickles the CORS issue.  If I get some time I'll try to clean up my existing cluster project to reproduce the issue.  The fix is so easy though it's hard to justify...

**Comment by slominskir** [comment]
Also worth pointing out it's not obvious how you did your test.  It appears you're running everything through the localhost loopback interface and varying ports to distinguish between hosts and then it's unclear how you're reaching host `admin.mykeycloak` from a browser, which presumably is the first Keycloak instance at 127.0.0.1:8443.   If you're launching all these servers manually I suppose you may have updated your local hosts file to add admin.mykeycloak.  If in Docker, then to reach admin.mykeycloak you'd need to run in docker host network mode, else also expose admin.mykeycloak to a localhost port to access it (which also changes the CORS situation).  It's not clear if you're on Windows. Mac, or Linux, and that may affect networking.

**Comment by slominskir** [comment]
Looks like cookie statement should include samesite=none, secure, and httponly (since re-encrypt used):

```
cookie KC_ROUTE insert indirect nocache secure httponly samesite=none
```

**Comment by slominskir** [comment]
The CORS issue on the admin console accessed via hostname-admin appears to be fixed in latest Keycloak (v21.0.1).  I was too lazy to create a reproducer docker compose for Keycloak v20.0.3, but I did roll out latest Keycloak to prod and noticed I didn't have to do an extra CORS fix.   Happy day.

**Comment by franck102** [comment]
I have spent a day 1/2 trying to setup keycloak in an AWS Fargate cluster, without success so far. I have read hundreds of posts about hostname, proxies, disabling ssl etc... so yes I agree that the documentation is really confusing.
Maybe a few suggestions of things that would help:

- Actual complete configuration example with the rationale for and effect of each setting explained

- Clearly categorizing config settings into those that affect the server request handling behavior, vs. those that affect URLs generated in the responses issued, vs. probably other categories that I am not aware of. A setting like "proxy: edge" is terrible because it leaves us guessing what may or may not be happening under the hood

- Clarifying the terminology in some instances: when I read "Keycloak is exposed through...", I don't know what that means for my keycloak container sitting behind an AWS load balancer that terminates SSL: does "expose" refer to how end-user see it (https://<some AWS CloudFront/ALD endpoint>) or does it refer to the local server network interfaces?

**Comment by vmuzikar** [comment]
I'm closing this in favor of https://github.com/keycloak/keycloak/issues/24252. Let me know if you have any concerns about it.

**Comment by shkmass11** [comment]
> I have spent a day 1/2 trying to setup keycloak in an AWS Fargate cluster, without success so far. I have read hundreds of posts about hostname, proxies, disabling ssl etc... so yes I agree that the documentation is really confusing. Maybe a few suggestions of things that would help:
> 
> * Actual complete configuration example with the rationale for and effect of each setting explained
> * Clearly categorizing config settings into those that affect the server request handling behavior, vs. those that affect URLs generated in the responses issued, vs. probably other categories that I am not aware of. A setting like "proxy: edge" is terrible because it leaves us guessing what may or may not be happening under the hood
> * Clarifying the terminology in some instances: when I read "Keycloak is exposed through...", I don't know what that means for my keycloak container sitting behind an AWS load balancer that terminates SSL: does "expose" refer to how end-user see it (https://<some AWS CloudFront/ALD endpoint>) or does it refer to the local server network interfaces?

Did you manage to get it up and running?

**Comment by franck102** [comment]
> Did you manage to get it up and running?

Yes I did, thanks!
